### PR TITLE
Delete empty paragraphs when backspacing

### LINKED ISSUE #3624: Backspace from empty paragraph into image doesn't delete paragraph
GIF:

backspace to delete empty block

Steps to reproduce:

- Have multiple paragraphs
- Backspace to empty out a paragraph and backspace into previous one, verify the empty paragraph disappears
- Now insert an image, and backspace into image from the empty paragraph block, and note that the paragraph doesn't disappear

### PR DESCRIPTION
**Pull request by noisysocks**
This is a rough attempt to fix #3624.

When backspacing from an empty paragraph into a block that has no merge function (e.g. an image), the empty paragraph should be deleted.

Out with the old:

old-behaviour

In with the new:

new-behaviour

### How to test

1. Have multiple paragraphs.
1. Backspace to empty out a paragraph and backspace into previous one, verify the empty paragraph disappears.
1. Now insert an image, and backspace into image from the empty paragraph block, verify that the empty paragraph disappears.

### CODE REVIEW

File: editor/effects.js [hunk 1]
```diff
@@ -217,6 +218,9 @@ export default {
 		// Only focus the previous block if it's not mergeable
 		if ( ! blockType.merge ) {
 			dispatch( focusBlock( blockA.uid ) );
+			if ( blockB.name === 'core/paragraph' && ! blockB.attributes.content.length ) {
```
Comment (noisysocks) [inline editor/effects.js]: My instinct here is to invent a `isBlockEmpty` helper which we would use instead of this very very specific logic. 

I'm not sure, though, since it probably _is_ only paragraphs that we want this behaviour for...

Thoughts, @aduth?

Comment (gziolo) [reply editor/effects.js]: We could also introduce a flag that marks all removable blocks like heading or list. 

`blockB.attributes.content.length` - is it safe to assume that content is there and it's a string?

Comment (gziolo) [reply editor/effects.js]: And yes, you can at least introduce a local variable which better explains what is happening in this conditional statement :)

**Comment by gziolo** [comment]
This `merge` property is also a bit strange, try the following:

Add an empty list and a quote and try to delete quote with backspace. It merges them together producing the list of 3 empty items:

empty-block

File: editor/effects.js (hunk 1 above)
Comment (youknowriad) [reply editor/effects.js]: Maybe we could introduce a ìsEmpty` property to the block API and if not defined the block is never empty. Only the block knows how to read its attributes.

Comment (noisysocks) [reply editor/effects.js]: I added an `isEmpty` property and implemented it for `core/paragraph`!

**Comment by noisysocks** [comment]
@gziolo it looks like that strangeness is caused by an unrelated bug in the `core/quote` → `core/list` transform. e5383ce473f4766671e59aa8572f265b966a24b8 ought to fix it.

**Comment by jasmussen** [comment]
Seriously cool work. In my testing, behavior wise, this works exactly like it needs to! 👍 👍 from an experience point of view.

File: editor/effects.js (hunk 1 above)
Comment (aduth) [reply editor/effects.js]: I don't love the `isEmpty` property (more to a general concern of limiting number of properties on the block API). It seems to me if Editable is smart enough to know that it should merge its own content into previous, it can also be smart enough to know that it has no content to merge and should just remove. That said, I also don't love an `onRemove` being added to Editable. Wondering if we could enhance one of the existing like `onReplace` to also encompass behavior of "replacing with nothing" ?

File: blocks/library/list/index.js [hunk 2]
```diff
@@ -118,8 +118,9 @@ registerBlockType( 'core/list', {
 				transform: ( { value, citation } ) => {
 					return createBlock( 'core/list', {
 						nodeName: 'UL',
-						values: value.map( ( item, index ) => <li key={ index } >{ get( item, 'children.props.children', '' ) } </li> )
-							.concat( citation ? <li key="citation">{ citation }</li> : [] ),
+						values: value.map( item => get( item, 'children.props.children' ) ).concat( citation )
+							.filter( text => typeof text === 'string' )
```
Comment (jorgefilipecosta) [inline blocks/library/list/index.js]: This was a nice find that we are creating a list with an empty <li> and this affects some functionality 👍
Unfortunately, we cannot be that strict and force the text to be a string. For example, if we have a quote with items in bold they are ignored in the transformation because these items are not a string.
Also if the user adds a quote with a line then 3 empty lines and another line, I'm not certain if we should just transform to a list with two items or to a list with one item 3 empty items and one last item.
Maybe we can just add a special case that verifies if value and citation are both empty/unset and if yes we just return with a list with no items. e.g: `if ( ( ! value || ! value.length ) && ! citation ) {`.

File: editor/effects.js (hunk 1 above)
Comment (noisysocks) [reply editor/effects.js]: That's a good idea @aduth. I moved my deletion logic to `Editable` in a0d193b7291da64bc9ce795f0f185e979b1bcbf9.

File: blocks/library/list/index.js (hunk 2 above)
Comment (noisysocks) [reply blocks/library/list/index.js]: I've fiddled with the logic in e2257f5dc42dd5c34c1bcce99d255f5a7fecb187. There's a surprising amount of edge cases in this logic. 

The logic I opted for is:

* `'Quote' / 'Citation'` → `<li>Quote</li><li>Citation</li>`
* `'' / 'Citation'` → `<li></li><li>Citation</li>`
* `'Quote' / ''` → `<li>Quote</li>`
* `'' / ''` → ``

**Comment by aduth** [comment]
A small thing that I really appreciate from this change: Clicking into the "Write your story" and hitting Backspace resets to remove the inserted paragraph block.

File: blocks/editable/index.js [hunk 3]
```diff
@@ -507,15 +507,27 @@ export default class Editable extends Component {
...
 			)
 		) {
-			const forward = event.keyCode === DELETE;
 			this.fireChange();
-			this.props.onMerge( forward );
+
+			const isEmpty = dom.isEmpty( rootNode );
+			if ( this.props.onReplace && isEmpty ) {
+				// Remove this block if the editor is empty
+				this.props.onReplace( [] );
```
Comment (aduth) [inline blocks/editable/index.js]: One odd thing about the logic flow here: If I render an `<Editable />` with `onReplace`, but not `onMerge`, I won't have this emptiness replace behavior, since this is only reached if `onMerge` is assigned, even though it's not called.

**Review by aduth — CHANGES_REQUESTED** [review]

**Comment by aduth** [comment]
I should note to my review, functionally it works great and I'm glad to see it handled in Editable entirely.

**Comment by aduth** [comment]
Hmm, a potential issue is that some blocks may have multiple Editables, and one of them being empty is not a sufficient metric to determine that the block should be replaced. Example: Pressing backspace from a quote citation, when the quote itself has text, shouldn't delete the block.

**Review by mcsf — CHANGES_REQUESTED** [review]
Nice work, _but_ it seems the replace behavior takes precedence over the merge one, leading to unexpected outcomes. Note how focus is lost when deleting into another paragraph:


| Before  | After |
| ------------- | ------------- |
| gutenberg-3732-merge-p | gutenberg-3732-replace-p |

**Comment by noisysocks** [comment]
> Hmm, a potential issue is that some blocks may have multiple Editables, and one of them being empty is not a sufficient metric to determine that the block should be replaced.

Good point... there's a lot of ~dragons~ subtleties here.

This approach _kind of_ works right now because in practice it's only `core/paragraph` that has both `onMerge` and `onReplace` passed to its `Editable`.

Unfortunately that means that the bug that this PR attempts to fix is only fixed for paragraph blocks and not quotes (which has `onMerge` but not `onReplace`):

quote-bug

If we give `core/quote` an `onReplace`, we fix the problem:

quote-bug-2

**BUT** doing this means we now have the problem you describe where a quote that contains only a citation can be incorrectly deleted:

quote-bug-3

---

Long story short, I think that patching `Editable` is unworkable. To fix this bug _in the general case_ we need the merge/deletion logic to read the block's entire state.

I have two ideas:

1) Go back to what we had before where each block defines an `isEmpty` method which the `MERGE_BLOCKS` effect can use to determine if that block should be deleted when being merged. I like the explicitness of this approach, but it _does_ mean extra API surface.
2) Same as (1) but instead of each block defining an `isEmpty` we build a method that compares each of the block's attributes against the block's default attributes. If no attribute differs from its default, then the block is empty and can be deleted when being merged. I think this might be our best bet.

**Comment by aduth** [comment]
Could we add a new callback `onRemove` to `Editable` which is called as in the current implementation where we delete an empty Editable? Then, the block could decide if `onRemove` should be handled as `onReplace( [] )` (i.e. paragraph, or in other blocks when first checking that all other fields are empty as well).

**Comment by noisysocks** [comment]
> Could we add a new callback onRemove to Editable which is called as in the current implementation where we delete an empty Editable? 

Nice suggestion. I've done this and it seems to work pretty well.

I also fixed a similar bug to what @gziolo found where if one presses backspace in an empty _paragraph_ block that follows a list block, an empty `<li>` is created.

File: blocks/editable/index.js [hunk 4]
```diff
@@ -507,17 +507,34 @@ export default class Editable extends Component {
...
+
+			const isEmpty = dom.isEmpty( rootNode );
+			if ( isEmpty && this.props.onRemove ) {
+				this.props.onRemove( forward );
+			}
+
 			event.preventDefault();
 			event.stopImmediatePropagation();
+
+			return;
```
Comment (aduth) [inline blocks/editable/index.js]: Is the `return;` necessary?

File: blocks/editable/index.js [hunk 5]
```diff
@@ -507,17 +507,34 @@ export default class Editable extends Component {
...
-			this.props.onMerge( forward );
+
+			const forward = event.keyCode === DELETE;
+
+			if ( this.props.onMerge ) {
+				this.props.onMerge( forward );
+			}
+
+			const isEmpty = dom.isEmpty( rootNode );
+			if ( isEmpty && this.props.onRemove ) {
```
Comment (aduth) [inline blocks/editable/index.js]: Minor: If `dom.isEmpty` is non-trivial ([sorta appears so](https://github.com/tinymce/tinymce/blob/master/src/core/src/main/js/dom/Empty.js)), we could achieve a minor optimization here with conditional short-circuiting by reordering the condition to test the truthy `onRemove` first:

```js
if ( this.props.onRemove && dom.isEmpty( rootNode ) ) {
```

**Review by aduth — APPROVED** [review]
This seems to work quite nicely.

**Comment by mtias** [comment]
Nice work here.
